lib/bsnutil/util.go

```diff
@@ -0,0 +1,202 @@
+	key := int(hashValue >> 1) // Shift right 1 bit to ensure positive int (31 bits)
+
+	return key
+}
+
+// generateRandomNonce creates a random nonce to make transport tokens unique.
+func generateRandomNonce() (string, error) {
+	// Generate 4 random bytes and convert to hex string
+	bytes := make([]byte, 4)
+	_, err := rand.Read(bytes)
+	if err != nil {
+		return "", fmt.Errorf("failed to generate random nonce: %w", err)
+	}
+	return fmt.Sprintf("%x", bytes), nil
+}
+
+// simpleXOR applies XOR transformation to the BSN using the generated key.
+// This is a placeholder for more sophisticated encryption in future versions.
+func simpleXOR(bsn, key int) int {
+	return bsn ^ key
+}
+
+// TODO: Remove this later - this logic will be implemented in a HAPI Interceptor at the NVI - only here to prove the concept
+// TransportTokenToPseudonym converts a transport token to a pseudonym format.
+// This extracts the core BSN information and creates a consistent pseudonym (ignoring timestamp/nonce).
+func TransportTokenToPseudonym(token string) (string, error) {
+	// Parse token components
+	audience, transformedBSN, err := parseTokenComponents(token)
+	if err != nil {
+		return "", fmt.Errorf("invalid token format")
+	}
+
+	// Generate consistent pseudonym using the transformed BSN and audience (deterministic)
+	return fmt.Sprintf("ps-%s-%d", audience, transformedBSN), nil
+}
+
+// TODO: Remove this later - this logic will be implemented in a HAPI Interceptor at the NVI - only here to prove the concept
+// PseudonymToTransportToken converts a pseudonym back to transport token format.
+// This reverses the TransportTokenToPseudonym transformation.
+func PseudonymToTransportToken(pseudonym string, audience string) (string, error) {
+	// Parse the pseudonym format manually to handle audience names with hyphens
+	if len(pseudonym) < minPseudonymLength || !strings.HasPrefix(pseudonym, pseudonymPrefix) {
+		return "", fmt.Errorf("invalid pseudonym format")
+	}
+
+	// Find the last hyphen to separate audience from transformedBSN
+	lastHyphen := strings.LastIndex(pseudonym[len(pseudonymPrefix):], "-")
+	if lastHyphen != -1 {
+		lastHyphen += len(pseudonymPrefix) // Adjust for the "ps-" prefix offset
+	}
+
+	if lastHyphen == -1 || lastHyphen == len(pseudonymPrefix) {
+		return "", fmt.Errorf("invalid pseudonym format")
+	}
+
+	pseudonymHolder := pseudonym[len(pseudonymPrefix):lastHyphen]
+	transformedBSNStr := pseudonym[lastHyphen+1:]
+
+	transformedBSN, err := strconv.Atoi(transformedBSNStr)
+	if err != nil {
+		return "", fmt.Errorf("invalid pseudonym format")
+	}
+
+	// Reverse the XOR to get original BSN
+	key := generateSimpleKey(pseudonymHolder)
+	originalBSN := simpleXOR(transformedBSN, key)
+
+	// Create new token with the target audience
+	return CreateTransportToken(originalBSN, audience)
+}
+
+// validateBSN checks if the BSN is within valid range for Dutch social security numbers.
+func validateBSN(bsn int) error {
+	if bsn < minBSN || bsn > maxBSN {
```

Comment:
I don't think we need to validate this here, since we can work with a simple plaintext, which can be whatever data (no need to check business rules here)